pkg/hubble/peer/buffer.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package peer
+
+import (
+	"fmt"
+	"io"
+
+	peerpb "github.com/cilium/cilium/api/v1/peer"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+type buffer struct {
+	max    int
+	buf    []*peerpb.ChangeNotification
+	mu     lock.Mutex
+	notify chan struct{}
+	stop   chan struct{}
+}
+
+// newBuffer creates a buffer of ChangeNotification that is safe for concurrent
+// use. The buffer is created with an initial size of 0 and is allowed to grow
+// until max is reached.
+func newBuffer(max int) *buffer {
+	return &buffer{
+		max:    max,
+		notify: nil,
+		stop:   make(chan struct{}),
+	}
+}
+
+// Len returns the number of elements in the buffer.
+func (b *buffer) Len() int {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return len(b.buf)
+}
+
+// Cap returns the capacity of the buffer.
+func (b *buffer) Cap() int {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return cap(b.buf)
+}
+
+// Push appends cn to the end of the buffer. An error is returned if its
+// maximum capacity is reached.
+func (b *buffer) Push(cn *peerpb.ChangeNotification) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if len(b.buf) == b.max {
+		return fmt.Errorf("max buffer size=%d reached", b.max)
+	}
+	b.buf = append(b.buf, cn)
+	if b.notify != nil {
+		close(b.notify)
+		b.notify = nil
+	}
+	return nil
+}
+
+// Pop removes and returns the first element in the buffer. If the buffer is
+// empty, Pop blocks until an element is added or Close is called in which case
+// io.EOF is returned.
+func (b *buffer) Pop() (*peerpb.ChangeNotification, error) {
+	b.mu.Lock()
+	if len(b.buf) == 0 {
+		if b.notify == nil {
+			b.notify = make(chan struct{})
+		}
+		notify := b.notify
+		b.mu.Unlock()
+		select {
+		case <-notify:
+			b.mu.Lock()
+		case <-b.stop:
+			return nil, io.EOF
+		}
+	}
+	var cn *peerpb.ChangeNotification
+	cn, b.buf = b.buf[0], b.buf[1:]
```

Comment:
nit:
```suggestion
 	cn := buf[0]
 	b.buf[0] = nil
 	b.buf = b.buf[1:]
```

otherwise `cn` will not be garabge collected until the `b.buf = append(b.buf, cn)` eventually drops the underlying buffer when reallocating.